### DISCUSSION TITLE
Overlap between kamal-proxy and thruster

### DISCUSSION THREAD

**Discussion by moonglum** [basecamp/kamal-proxy#55] [Q&A] [answered]
At his talk at Rails World, @kevinmcconnell presented these new features for kamal-proxy:
![Slide of the new features of kamal-proxy](https://github.com/user-attachments/assets/514dfca0-4fbb-4a1f-bb7c-d51bf06f3e6f)

In his presentation, he states that it is intended to be used together with thruster (which he also co-created). I’m slightly confused by this, as thruster has a subset of the features of kamal-proxy. From the [introduction post of thruster](https://dev.37signals.com/thruster-released/), its features are:

- HTTP/2 support
- TLS (via [Let’s Encrypt](https://letsencrypt.org/), for automatic certificate management)
- HTTP caching, for public assets
- Efficient static file serving via X-Sendfile and compression

This is a subset of the features above. Is this a misunderstanding from my side that some of the features on the slide are features in thruster not kamal-proxy?

**Comment by kevinmcconnell** [comment, marked as answer]
Hi @moonglum,

I agree it's a bit confusing :) The HTTP caching & X-Sendfile features are provided by Thruster. I tried to mention that in the talk, but may not have made it entirely clear. Those features are handled inside Thruster because it simplifies some aspects of them to be inside the app container. For example, X-Sendfile works by having the application return paths to locations on disk where a file should be served from, so whatever does that serving needs to have access to those paths. Handling that in another container would require configuring some sort of shared volume, whereas running it inside the app container means it can work without any configuration. It's simpler that way. So we use Thruster to add these extra acceleration features, while Kamal Proxy is responsible for serving and routing traffic into the containers.

The other part that may be confusing is that Thruster can also be used on its own, without any sort of proxy layer (if you don't need gapless deployments or other Kamal features). That's how the ONCE apps work. Because of that, Thruster can also do the TLS, HTTP/2 parts if necessary. But when running behind Kamal Proxy it won't need to do that, because Kamal Proxy will already have taken care of it. So there is a bit of overlap in features there because of the flexibility in how Thruster can be used.

Hope that helps clarify things!

**Reply by moonglum** [reply]
Thanks a lot :) That makes sense to me. Let me rephrase it in my own words to make sure I understand it correctly.

There are basically three different setups:

**Single Server with Thruster only:**
* Thruster takes care of TLS (via Let’s Encrypt) and HTTP/2
* Thruster does HTTP caching
* Thruster handles sending static files via X-Sendfile
* (The rest is handled by puma/rails)

**Single Server with kamal-proxy and thruster:**
* kamal-proxy shifts traffic from old instance to new instance of an app
* Kamal-proxy gives you the option to have multiple apps on the server
* kamal-proxy offers maintenance pages, request pausing and canary deployments
* kamal-proxy buffers requests/responses
* kamal-proxy takes care of TLS (via Let’s Encrypt) and HTTP/2
* Thruster does HTTP caching
* Thruster handles sending static files via X-Sendfile
* (The rest is handled by puma/rails)

**Multiple Servers with kamal-proxy and thruster:**
* Load balancing is not handled by kamal, so it needs to be done by the ops person
* In most cases, TLS and H/2 will be done by the load balancer as well (unless layer 4 load balancing is used - but I’m not sure the Let’s Encrypt setup in kamal-proxy would work in that case)
* Maintenance Pages/Request Pausing/Canary Deployments could be used in theory, but the ops person would need to do it for each of the servers
* kamal-proxy buffers requests/responses
* Thruster does HTTP caching
* Thruster handles sending static files via X-Sendfile
* (The rest is handled by puma/rails)

**Reply by kevinmcconnell** [reply]
Yes, although I'd just mention a couple points for the "multiple servers" case:

- Kamal Proxy is still handling the traffic shift between old & new instances in that case, same as with the singe server setup
- It also still allows you to run multiple apps per server (e.g. maybe you have 5 apps, using a pool of 3 servers that they all run across)
- You could use layer 4 load balancing if you supply your own certs to the proxy (we don't currently support auto TLS over multiple servers at once, although we might in the future, but we do allow you to configure your own certs)
- Features like maintenance pages & pausing will be available as Kamal commands in the near future. So in that setup, you will execute a command like "pause" once, and Kamal will handle coordinating that across all your servers for you. These features are very much intended to be useful in a multi-server setup, it's just that we haven't finished exposing them in Kamal quite yet.

**Reply by kevinmcconnell** [reply]
Oh and I'm working on adding load balancing to Kamal as well, which should simplify a lot of that :) Don't have an ETA on it yet though.

**Reply by moonglum** [reply]
Excellent! Thanks a lot for the clarifications and taking the time, Kevin! 🙇
